_posts/2024-06-14-authz-openfga-introduction.md

```diff
@@ -0,0 +1,264 @@
+
+By specifying types like `user` and `document`, and defining possible relationships such as `reader`, `writer`, and `owner`, the model establishes who can do what and under what circumstances. 
+Each type can be seen as a category of objects, and each relationship defines how these objects can interact, ensuring that permissions are granted appropriately and securely.
+
+Here’s how an authorization model is expressed in OpenFGA’s DSL, which highlights its role in setting up the system’s permission framework:
+
+```
+model:
+    schema: 1.1
+
+type: user
+
+type: document
+    relations:
+        define reader as [user]
+        define writer as [user]
+        define owner as [user]
+```
+
+## Testing Capabilities in OpenFGA
+
+OpenFGA's testing capabilities offer a significant advantage by allowing teams to validate their authorization models before integration into live applications.
+This process not only enhances security and compliance but also ensures that the authorization logic performs as expected across different scenarios.
+
+### Recap of the Scenario with Jane and GlobeCorp
+
+Previously, we discussed the scenario involving Jane, an auditor at GlobeCorp. 
+Jane had 'readonly' access to GlobeCorp's financial data and also held an admin role at EcoLife. 
+This scenario sets the stage for testing complex permission structures to ensure they function correctly within the system.
+
+![Relations](/assets/article_images/2024-06-14-authz-openfga-introduction/global-corp-2.png)
+
+### Defining the Authorization Model
+
+To encapsulate Jane's permissions within GlobeCorp, we crafted an authorization model. Here's how it is defined in OpenFGA's format:
+
+```
+model:
+  schema: 1.1
+
+type: user
+
+type: organization
+  relations:
+    define owner as [user] or owner from parent
+    define parent as [organization]
+    define editor as [user] or editor from parent or owner
+    define viewer as [user] or viewer from parent or owner
+
+    define can_view as viewer
+    define can_edit as editor
+```
+
+This model specifies hierarchical permissions, reflecting Jane's access levels across different organizational layers.
+
+### Testing the Model
+
+Tests in OpenFGA are specified in a YAML format, which makes them easily readable and maintainable. 
+For Jane's scenario, we could define a test to verify that she has the correct `can_view` permissions for GlobeCorp's data and the `can_edit` permissions for EcoLife.
+
+Here’s an example of how such a test might look:
+
+```yaml
+name: demo
+model_file: ./model.fga
+# Sets up the initial state by defining specific relationships
+tuples:
+  - user: organization:GlobalCorp
+    relation: parent
+    object: organization:EcoLIfe
+  - user: user:Jane
+    relation: viewer
+    object: organization:GlobalCorp
+  - user: user:Jane
+    relation: editor
+    object:  organization:EcoLIfe
```

Comment:
question aside, why doesn't this use the special syntax explained earlier, `organization:EcoLife#editor@Jane`?

(btw EcoLife `I` should be small probably)

Reply:
This is how they define the DSL. I would say it is easier to write tests instead to use the academic zansibar syntax.